README.md
# NAME

OrePAN2::Server - DarkPAN Server

# SYNOPSIS

    #launch orepan2 standalone server http://localhost:5888/
    % orepan2-server.pl -p 5888

    #upload git managed module to my orepan2 by curl 
    curl --data-urlencode 'module=git@github.com:Songmu/p5-App-RunCron.git' --data-urlencode 'author=SONGMU' http://localhost:5888/
    curl --data-urlencode 'module=git+ssh://git@mygit/home/git/repos/MyModule.git' --data-urlencode 'author=SONGMU' http://localhost:5888/
    curl --data-urlencode 'module=git+file:///home/hirobanex/project/MyModule.git' --data-urlencode 'author=SONGMU' http://localhost:5888/

    #install by cpanm
    cpanm --mirror=http://localhost:5888/orepan Your::Module

    #install by carton install
    PERL_CARTON_MIRROR=http://localhost:5888/orepan carton install

# DESCRIPTION

OrePAN2::Server is DarkPAN server, or [OrePAN2](http://search.cpan.org/perldoc?OrePAN2) Uploader that use API provided by OrePAN2.

Like uploading to cpan, you can upload to your DarkPAN by http post request.

If you set your DarkPAN url in options([cpanm](http://search.cpan.org/perldoc?cpanm) --mirror, [carton](http://search.cpan.org/perldoc?carton)  PERL\_CARTON\_MIRROR), you can easily install and manage your modules in your project.

You should set up DarkPAN in private space. If you upload your modules to DarkPAN on public space, you consider to upload your modules to cpan. 

# USAGE

## launch OrePAN2 server instantly

See [orepan2-server.pl](http://search.cpan.org/perldoc?orepan2-server.pl)

## attach your plack app.

    use Plack::Builder;
    use OrePAN2::Server::CLI;
    use Your::App;

    my $orepan = OrePAN2::Server::CLI->new_with_options(
        delivery_dir     => "orepan",
        delivery_path    => "/",
        authenquery_path => "/authenquery",
        compress_index   => 1,
    );

    builder {
        mount '/'       => Your::App->to_app();
        mount '/orepan' => $orepan->app;
    };

If your need only DarkPAN Uploader and add Basic Auth with `Plack::Middleware::Auth::Basic`, you code this.

    use Plack::Builder;
    use OrePAN2::Server;
    use Your::App;

    my $orepan_uploader = OrePAN2::Server->uploader(
        directory        => "orepan",
        compress_index   => 1,
    );

    builder {
        mount '/'            => Your::App->to_app();
        mount '/authenquery' => builder {
            enable "Auth::Basic", authenticator => sub { return ($_[0] eq 'userid' && $_[1] eq 'password') };
            $orepan_uploader;
        }
    };



## upload by minil release.

There is three step.

### minil.toml

    [release]
    pause_config="/path/to/your-module/.pause"

### /path/to/your-module/.pause

    upload_uri http://orepan2-server/authenquery
    user hirobanex
    password dummy

password is dummy.this feature is not officially documented in [CPAN::Uploader](http://search.cpan.org/perldoc?CPAN::Uploader).See [Minilla](http://search.cpan.org/perldoc?Minilla).

### upload command

minil release

# SEE ALSO

[orepan2-server.pl](http://search.cpan.org/perldoc?orepan2-server.pl), [OrePAN2](http://search.cpan.org/perldoc?OrePAN2), [Minilla](http://search.cpan.org/perldoc?Minilla)

# LICENSE

Copyright (C) Hiroyuki Akabane.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Hiroyuki Akabane <hirobanex@gmail.com>

Songmu <y.songmu@gmail.com>


Changes
Revision history for Perl extension OrePAN2-Server

{{$NEXT}}

0.04 2013-11-03T16:23:11Z
    - modify document that CPAN::Uploader document support  upload uri

0.03 2013-11-03T15:03:20Z
    - return 405 against to access /authenquery by none post request
    - use local (not remote) git repo in test
    - maintainance document

0.02 2013-11-01T09:01:37Z
    - fix typo in document 

0.01 2013-11-01T08:51:22Z
    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'OrePAN2-Server',
    module_name     => 'OrePAN2::Server',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/OrePAN2/Server/CLI.pm
package OrePAN2::Server::CLI;
use strict;
use warnings;
use utf8;

use File::Path ();
use Plack::App::Directory;
use Plack::Builder;
use Plack::Runner;

use OrePAN2::Server;

use Getopt::Long qw/:config posix_default no_ignore_case bundling auto_help pass_through/;
use Pod::Usage   qw/pod2usage/;

sub new {
    my $class = shift;
    local @ARGV = @_;

    my %opt = ('compress-index' => 1);
    GetOptions(\%opt, qw/
        delivery-dir=s
        delivery-path=s
        authenquery-path=s
        compress-index!
    /) or pod2usage(1);

    $opt{delivery_dir}     = delete $opt{'delivery-dir'}     || 'orepan';
    $opt{delivery_path}    = delete $opt{'delivery-path'}    || $opt{delivery_dir};
    $opt{authenquery_path} = delete $opt{'authenquery-path'} || 'authenquery';
    $opt{compress_index}   = delete $opt{'compress-index'};

    for my $path (qw/delivery_path authenquery_path/) {
        $opt{$path} = '/' . $opt{$path} if $opt{$path} !~ m!^/!;
    }
    $opt{argv} = \@ARGV;

    $class->new_with_options(%opt);
}

sub new_with_options {
    my $class = shift;
    bless {@_}, $class;
}

sub app {
    my $self = shift;

    $self->{app} ||= do {
        File::Path::mkpath $self->{delivery_dir};
        my $app = builder {
            mount $self->{authenquery_path} => OrePAN2::Server->uploader(
                directory      => $self->{delivery_dir},
                compress_index => $self->{compress_index},
            );
            mount $self->{delivery_path} => Plack::App::Directory->new({root=> $self->{delivery_dir}})->to_app;
        };
    };
}

sub run {
    my $self = shift;

    my $runner = Plack::Runner->new;
    $runner->parse_options(@{ $self->{argv} || [] });
    $runner->run($self->app);
}

1;


lib/OrePAN2/Server.pm
package OrePAN2::Server;
use 5.008001;
use strict;
use warnings;

our $VERSION = "0.03";

use File::Copy ();
use File::Spec;
use File::Temp ();
use Plack::Request;
use OrePAN2::Injector;
use OrePAN2::Indexer;

sub uploader {
    my ($class, %args) = @_;
    my $directory      = $args{directory} || 'orepan';
    my $compress_index = 1;
    $compress_index = $args{compress_index} if exists $args{compress_index};

    return sub {
        my $env = shift;
        my $req = Plack::Request->new($env);
        return [404, [], ['NOT FOUND']] if $req->path_info !~ m!\A/?\z!ms;

        if ($req->method eq 'POST') {
            eval {
                my ($module, $author);

                my $tempdir = File::Temp::tempdir( CLEANUP => 1 );
                if (my $upload = $req->upload('pause99_add_uri_httpupload')) {
                    # request from CPAN::Uploader
                    $module = File::Spec->catfile($tempdir, $upload->filename);
                    File::Copy::move $upload->tempname, $module;
                    $author = $req->param('HIDDENNAME');
                }
                else {
                    $module = $req->param('module'); # can be a git repo.
                    $author = $req->param('author') || 'DUMMY';
                }
                return [404, [], ['NOT FOUND']] if !$module && !$author;
                $author = uc $author;

                my $injector = OrePAN2::Injector->new(
                    directory => $directory,
                    author    => $author,
                );
                $injector->inject($module);

                OrePAN2::Indexer->new(directory => $directory)->make_index(
                    no_compress => !$compress_index,
                );
            };

            if (my $err = $@) {
                warn $err . '';
                return [500, [], [$err.'']];
            }
        }
        else {
            return [405, [], ['Method Not Allowed']];
        }

        return [200, [], ['OK']];
    }
}


1;
__END__

=encoding utf-8

=head1 NAME

OrePAN2::Server - DarkPAN Server

=head1 SYNOPSIS

    #launch orepan2 standalone server http://localhost:5888/
    % orepan2-server.pl -p 5888

    #upload git managed module to my orepan2 by curl 
    curl --data-urlencode 'module=git@github.com:Songmu/p5-App-RunCron.git' --data-urlencode 'author=SONGMU' http://localhost:5888/
    curl --data-urlencode 'module=git+ssh://git@mygit/home/git/repos/MyModule.git' --data-urlencode 'author=SONGMU' http://localhost:5888/
    curl --data-urlencode 'module=git+file:///home/hirobanex/project/MyModule.git' --data-urlencode 'author=SONGMU' http://localhost:5888/

    #install by cpanm
    cpanm --mirror=http://localhost:5888/orepan Your::Module

    #install by carton install
    PERL_CARTON_MIRROR=http://localhost:5888/orepan carton install

=head1 DESCRIPTION

OrePAN2::Server is DarkPAN server, or L<OrePAN2> Uploader that use API provided by OrePAN2.

Like uploading to cpan, you can upload to your DarkPAN by http post request.

If you set your DarkPAN url in options(L<cpanm> --mirror, L<carton>  PERL_CARTON_MIRROR), you can easily install and manage your modules in your project.

You should set up DarkPAN in private space. If you upload your modules to DarkPAN on public space, you consider to upload your modules to cpan. 

=head1 USAGE

=head2 launch OrePAN2 server instantly

See L<orepan2-server.pl>

=head2 attach your plack app.

    use Plack::Builder;
    use OrePAN2::Server::CLI;
    use Your::App;

    my $orepan = OrePAN2::Server::CLI->new_with_options(
        delivery_dir     => "orepan",
        delivery_path    => "/",
        authenquery_path => "/authenquery",
        compress_index   => 1,
    );

    builder {
        mount '/'       => Your::App->to_app();
        mount '/orepan' => $orepan->app;
    };

=head2 attach your plack with Basic Auth

If your need only DarkPAN Uploader and add Basic Auth with C<Plack::Middleware::Auth::Basic>, you code this.

    use Plack::Builder;
    use OrePAN2::Server;
    use Your::App;

    my $orepan_uploader = OrePAN2::Server->uploader(
        directory        => "orepan",
        compress_index   => 1,
    );

    builder {
        mount '/'            => Your::App->to_app();
        mount '/authenquery' => builder {
            enable "Auth::Basic", authenticator => sub { return ($_[0] eq 'userid' && $_[1] eq 'password') };
            $orepan_uploader;
        }
    };


=head2 upload by minil release.

There is three step.

=head3 minil.toml

    [release]
    pause_config="/path/to/your-module/.pause"

If you want to know other options,  See L<Minilla>.

=head3 /path/to/your-module/.pause

    upload_uri http://orepan2-server/authenquery
    user hirobanex
    password password

If you want to know other options,  See L<CPAN::Uploader>.

You must pay attention to set your DarkPAN uri as upload_uri.If you don't, you will upload to cpan!

=head3 upload command

minil release

=head1 SEE ALSO

L<orepan2-server.pl>, L<OrePAN2>, L<Minilla>

=head1 LICENSE

Copyright (C) Hiroyuki Akabane.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Hiroyuki Akabane E<lt>hirobanex@gmail.comE<gt>

Songmu E<lt>y.songmu@gmail.comE<gt>

=for stopwords OrePAN DarkPAN

=cut

